### DISCUSSION TITLE
NanoEls H4 Feedback

### DISCUSSION THREAD

**Discussion by fmw626** [kachurovskiy/nanoels#124] [Q&A]
Hi everyone,

I am writing this post to share my feedback, observations, and questions following the completion of [my NanoEls H4](https://github.com/kachurovskiy/nanoels/discussions/123) build. I'd like to reiterate that the H4 is an awesome piece of kit as it is, and this post is meant to be a discussion about my personal observations!

@kachurovskiy, what steppers and settings are you using? I am struggling to find a good balance between current, heat and H4 parameters. Somehow after running for a while, especially the X axis starts acting very weirdly, like it accelerates slowly at first, then very quickly and stalls. Even after tweaking speed/acceleration settings, this behavior persists:

https://github.com/kachurovskiy/nanoels/assets/122809499/79672941-1548-4521-94f5-af413cbe38d5 

(how do you embed a video player into a post anyway?)

this is my hardware: 
Z: Nema 24 Closed Loop + CL57T @48v 1:1 gear ratio (4Nm) 4 Microsteps (800steps/rev)
X: Nema 17 Open Loop + DM542T @48v 3:1 gear ratio (1,95Nm) 4 Microsteps (800steps/rev)

these are my settings:
```
// Main lead screw (Z) parameters.
#define MOTOR_STEPS_Z 800.0 // 800
#define SCREW_Z_DU 20000.0 // 2mm lead screw in deci-microns (10^-7 of a meter)
#define SPEED_START_Z (2 * MOTOR_STEPS_Z) // Initial speed of a motor, steps / second.
#define ACCELERATION_Z (20 * MOTOR_STEPS_Z) // Acceleration of a motor, steps / second ^ 2.
#define SPEED_MANUAL_MOVE_Z (5 * MOTOR_STEPS_Z) // Maximum speed of a motor during manual move, steps / second.
#define INVERT_Z true // change (true/false) if the carriage moves e.g. "left" when you press "right".
#define NEEDS_REST_Z false // Set to false for closed-loop drivers, true for open-loop.
#define MAX_TRAVEL_MM_Z 400 // Lathe bed doesn't allow to travel more than this in one go, 30cm / ~1 foot
#define BACKLASH_DU_Z 2700 // 6500 = 0.65mm backlash in deci-microns (10^-7 of a meter) - FMW edit to 2700 (0,27mm)

// Cross-slide lead screw (X) parameters.
#define MOTOR_STEPS_X 2400.0 // 800 steps at 3x reduction
#define SCREW_X_DU 10000.0 // 1.25mm lead screw with 3x reduction in deci-microns (10^-7) of a meter - FMW edit to 10000 because 1mm leadscrew
#define SPEED_START_X (MOTOR_STEPS_X) // Initial speed of a motor, steps / second.
#define ACCELERATION_X (5 * MOTOR_STEPS_X) // Acceleration of a motor, steps / second ^ 2.
#define SPEED_MANUAL_MOVE_X (5 * MOTOR_STEPS_X) // Maximum speed of a motor during manual move, steps / second.
#define INVERT_X false // change (true/false) if the carriage moves e.g. "left" when you press "right".
#define NEEDS_REST_X false // Set to false for all kinds of drivers or X will be unlocked when not moving.
#define MAX_TRAVEL_MM_X 120 // Cross slide doesn't allow to travel more than this in one go, 10cm
#define BACKLASH_DU_X 1150 // 1500 = 0.15mm backlash in deci-microns (10^-7 of a meter) - FMW edit to 1150 (0,115mm)
```

Here is a list of behaviors that felt unintuitive to me. It could very well be that some of these are down to user error 🙃  :
- when long-pressing X move buttons, the movement doesn't stay within the selected steps, whereas it does on the Z axis. The Z axis behavior makes more sense to me. If your step is set to 1mm, it makes sense to move in 1mm increments both on Z an X no?
- like you have [posted before](https://github.com/kachurovskiy/nanoels/discussions/120#discussioncomment-6082502): the current position display disappears when the current position is 0. Also, it would make more sense to me to have the position display always as such:
     Z=10.000mm   X=10.000mm
- like you have suggested [here](https://github.com/kachurovskiy/nanoels/discussions/121#discussioncomment-6083895), I think that arrow symbols would be more intuitive than "LRUD"
- for the automated multi-operation features:
     - being asked the number of passes is already great, but even more useful would be to set the depth of cut instead. I manually calculate the number of passes depending on the desired depth of cut. The number of passes could be easily calculated from the depth of cut, and would need to be set less often in my opinion.
- Cone mode: the behavior different than facing/turning. It doesn't do multiple passes? Would it not make sense to face from a corner of the LRUD rectangle towards the diagonal? This is the most intuitive movement I can think of: ![Screenshot 2023-06-12 at 21 49 04](https://github.com/kachurovskiy/nanoels/assets/122809499/de059f34-edfe-4a5e-8fd7-01116ce7b380) 
     Also, I'm getting unexpected travel to the left when pressing right at the end of the cone cut!! 
- Cutoff: Why ask the number of passes? Shouldn't it always be only one pass?
- Lathe running in reverse behavior or on screen warning? As of now, the NanoEls just does… nothing.
- Facing operations: facing Up to Down as well as Down to Up (or in reverse). For example, when using a boring bar to face a relief. I ended up doing an internal turning operation instead.

Additional features I thought of:
- jump to next operation: map one button to jump to the next operation. Maybe ![239652261-1bf41a47-f4b5-4896-96a6-72c97ae472e2](https://github.com/kachurovskiy/nanoels/assets/122809499/e6aaaa36-aa73-412d-9647-ee938032ba0f)? Could be very useful if the soft stops were set hastily and the tool doesn't engage the workpiece yet.
- Quick Jog / precision mode toggle? (going through 1mm 0,1mm 0,01mm 0,001mm every time gets old quick 😃 )
- Set current X/Y position to XXX (as opposed to reset current position to 0 only)
- Possibility to disable/enable all soft stops (for example to move the carriage out of the way to measure the part, without losing the stop positions)

What do you think?

**Comment by kachurovskiy** [comment]
Hi @fmw626, thanks a lot for the feedback, I will be taking a close look at every item shortly. For now can you please provide more info on bugs - motor stall and unexpected movement after cone?

You can embed videos under 10 megabytes by drag and dropping them into the comment editor. Thanks a lot 🙏

**Reply by fmw626** [reply]
Hi @kachurovskiy 
here are two clips showing the unexpected movement after cone. Please watch them with sound:

https://github.com/kachurovskiy/nanoels/assets/122809499/37e14045-7a73-4a10-8fb3-b39e303610b8


https://github.com/kachurovskiy/nanoels/assets/122809499/f890c65d-48bc-4c8a-b612-4428ea56055e

The LRUD working area is as follows: LR=20mm, UD=10mm. The start of the cone is at the RU point, cone ratio is 2 (45°). It cuts the cone just fine. After the tool comes to a stop, if i press ![239697428-48a0327e-0cc4-465e-b371-644f153f3288](https://github.com/kachurovskiy/nanoels/assets/122809499/3b1845f0-a104-4c0a-9231-445c21899f33), the tool moves slightly to the left first before moving to the right, which could cause a crash with the workpiece. The amount of left movement seems to be random.

**Reply by kachurovskiy** [reply]
Thank you, I was finally able to reproduce it. I have the fixed version of the code available in https://gist.github.com/kachurovskiy/d69f8575cdb931938e9f5d28bb52d942 but since it changes more than a few lines I'm going to do some further testing before committing it to the main repo.

**Reply by fmw626** [reply]
Awesome, thank you @kachurovskiy. I will give the fixed code a try.

**Reply by kachurovskiy** [reply]
The fix is now in the main repo - https://github.com/kachurovskiy/nanoels/commit/2134a98fb0e6f22e5bdb20405a8e6dc0a77ec2c3 - just download the latest h4.ino to try - thanks a lot!

**Comment by fmw626** [comment]
Hi @kachurovskiy
here are four clips showing the X stepper stall. Please watch them with sound.

Normal behavior:

https://github.com/kachurovskiy/nanoels/assets/122809499/2551b7e5-1fda-4ea6-9253-1408055fad09


https://github.com/kachurovskiy/nanoels/assets/122809499/0ea1c5b8-81e1-465e-a17b-40dcdc05ea2d



Stalling behavior:

https://github.com/kachurovskiy/nanoels/assets/122809499/5c34846a-034e-487b-b50c-a809be190082


https://github.com/kachurovskiy/nanoels/assets/122809499/71901561-31f3-4995-878d-5549bc433165

The stalling behavior starts after a while running the H4, roughly 10-20 min or so. It can happen during manual movement or during automatic operations. Since it looses steps, it also loses its position and can scrap the part. It seems to me like it accelerates too slowly, then very abruptly, ending in a stall. You can see when you scrub the video frame by frame that when the stepper sound pitches up, the amount of rotation in the last frame before the stall is more than before, so it briefly sped up a lot. I cannot say as of now if the trigger is power-on time or amount traveled. The only way to resolve the issue is to power cycle the H4 (no need to power cycle the steppers). Dissabling and enabling the X stepper via the ![239705610-3059b6ed-0197-4e48-91a7-80a7e1317176](https://github.com/kachurovskiy/nanoels/assets/122809499/9a2055af-238f-4f84-be0d-ac526177a18c) button on the H4 doesnt solve the issue.

**Reply by fmw626** [reply]
> I cannot say as of now if the trigger is power-on time or amount traveled. The only way to resolve the issue is to power cycle the H4 (no need to power cycle the steppers). Dissabling and enabling the X stepper via the 239705610-3059b6ed-0197-4e48-91a7-80a7e1317176 button on the H4 doesnt solve the issue.

After further testing, i can confirm that the issue does not arise from power-on time. I left the H4 and steppers powered on for hours without touching the controls, and the stalling behavior did not occur.

**Comment by kachurovskiy** [comment]
> what steppers and settings are you using?

Almost the same as you, you can see my settings in https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino - since you mentioned that rebooting ELS fixes the issue without rebooting the stepper drivers I suspect they are not at fault.

Your settings look good.

> when long-pressing X move buttons, the movement doesn't stay within the selected steps, whereas it does on the Z axis. The Z axis behavior makes more sense to me. If your step is set to 1mm, it makes sense to move in 1mm increments both on Z an X no?

The only case when it doesn't follow the step is when you request to travel beyond the limit - in that case it stops on the limit - and when you travel back, you end up on a different point from where you started. Please clarify of post a short video to show what you mean, sorry if I'm missing something.

> Also, it would make more sense to me to have the position display always as such: Z=10.000mm X=10.000mm

We only have 20 chars to work with, position can be negative and can be 100 or more mm - in this case, `Z=-100.000mm X=-100.000mm` is 25 characters. `-100.000mm` is exactly 10 characters. Maybe we should make a special character for `mm` that takes 1 char - something like ![image](https://github.com/kachurovskiy/nanoels/assets/517919/7d143dea-0cc8-449a-a61b-e123ba5c422e).

I hide X=0 because some people would use H4 with just 1 axis but I agree that repeated feedback shows that folks want to see that 0 and I should make it possible to show it.

> being asked the number of passes is already great, but even more useful would be to set the depth of cut instead. I manually calculate the number of passes depending on the desired depth of cut. The number of passes could be easily calculated from the depth of cut, and would need to be set less often in my opinion.

Entering the DOC is much more clicks though? I can see the value in automatically calculating the number of passes though based on some "default" DOC. Currently number of passes is saved and reused each time, perhaps that's not ideal, need to think about it and collect more feedback.

> Cone mode: the behavior different than facing/turning. It doesn't do multiple passes? Would it not make sense to face from a corner of the LRUD rectangle towards the diagonal?

Yes, it's different - mostly because I figured that it's going to be pretty hard for someone to calculate the values for stops - so they can't define the cone ratio. If the ratio is separate from stops, then it becomes pretty complicated UI-wise - too many ways to setup a cone operation. With current approach you can easily cut cone left-right-left-right without having to move back - just by using reverse button. You also don't have to wrestle with calculations much. I think the ultimate answer here would be https://kachurovskiy.github.io/lathecode/ which (once operational) will take all the guesswork out.

> Also, I'm getting unexpected travel to the left when pressing right at the end of the cone cut!!

This is pretty scary, tried to reproduce but unable so far, will keep trying. We'll fix it.

> Lathe running in reverse behavior or on screen warning? As of now, the NanoEls just does… nothing.

You mean in automatic mode? Yes, an on-screen warning with a beep would be nice. Hopefully not a pressing issue :)

> Facing operations: facing Up to Down as well as Down to Up (or in reverse). For example, when using a boring bar to face a relief. I ended up doing an internal turning operation instead.

I believe this is supported already, just set the negative pitch.

> jump to next operation: map one button to jump to the next operation. Could be very useful if the soft stops were set hastily and the tool doesn't engage the workpiece yet.

This is an interesting idea and probably not hard to implement :) I'll see if it works.

> Quick Jog / precision mode toggle? (going through 1mm 0,1mm 0,01mm 0,001mm every time gets old quick 😃 )

No idea how that would work UI-wise.

> Set current X/Y position to XXX (as opposed to reset current position to 0 only)

This is high on my priority list, it's needed for GCode zeroing.

> Possibility to disable/enable all soft stops (for example to move the carriage out of the way to measure the part, without losing the stop positions)

I was thinking about it but don't see a practical necessity for it yet, setting a right stop in a slightly different place after measuring doesn't seem to affect any of the operations AFAIU?

Huge thanks for taking time to write this feedback 🙏

**Reply by fmw626** [reply]
>>  what steppers and settings are you using?

> Almost the same as you, you can see my settings in https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino - since you mentioned that rebooting ELS fixes the issue without rebooting the stepper drivers I suspect they are not at fault.

> Your settings look good.

👍👍



> > when long-pressing X move buttons, the movement doesn't stay within the selected steps, whereas it does on the Z axis. The Z axis behavior makes more sense to me. If your step is set to 1mm, it makes sense to move in 1mm increments both on Z an X no?

> The only case when it doesn't follow the step is when you request to travel beyond the limit - in that case it stops on the limit - and when you travel back, you end up on a different point from where you started. Please clarify of post a short video to show what you mean, sorry if I'm missing something.

https://github.com/kachurovskiy/nanoels/assets/122809499/8859cf41-c128-46fd-afaa-912a8c8339c9

This is the behavior I am talking about. No soft stops. Z stays neatly within the selected steps, regardless if i do a single click (1 step) or keep the button pressed (x  full steps). On the X axis, one single click = 1 step, but when i keep the button pressed, the movement disregards the steps. It would, in my opinion, make more sense if the steps are kept for all 4 described situations. 
That a limit disregards the step is totally fine and expected.



>>  Also, it would make more sense to me to have the position display always as such: Z=10.000mm X=10.000mm

> We only have 20 chars to work with, position can be negative and can be 100 or more mm - in this case, Z=-100.000mm X=-100.000mm is 25 characters. -100.000mm is exactly 10 characters. Maybe we should make a special character for mm that takes 1 char - something like image.

Ah, of course! I haven't thought of character limits. Since the LCD characters are 5x7 px, what do you think of this 1 character "mm" design:
![Screenshot 2023-06-13 at 22 14 49](https://github.com/kachurovskiy/nanoels/assets/122809499/f9ad28fe-9c6a-4d77-8d1f-15a78fb99d14)

> I hide X=0 because some people would use H4 with just 1 axis but I agree that repeated feedback shows that folks want to see that 0 and I should make it possible to show it.

👍👍



>>  being asked the number of passes is already great, but even more useful would be to set the depth of cut instead. I manually calculate the number of passes depending on the desired depth of cut. The number of passes could be easily calculated from the depth of cut, and would need to be set less often in my opinion.

> Entering the DOC is much more clicks though? I can see the value in automatically calculating the number of passes though based on some "default" DOC. Currently number of passes is saved and reused each time, perhaps that's not ideal, need to think about it and collect more feedback.

I believe that DOC would still make more sense. It might be more clicks, but with the number buttons it's quite quick. Besides, the depth of cut would generally be reused for several automated jobs, so you would probably end up setting it less often than the number of passes. Don't you think? Having to do the math from DOC to NOP before every job is definitely more tedious than entering the DOC.



>>  Cone mode: the behavior different than facing/turning. It doesn't do multiple passes? Would it not make sense to face from a corner of the LRUD rectangle towards the diagonal?

> Yes, it's different - mostly because I figured that it's going to be pretty hard for someone to calculate the values for stops - so they can't define the cone ratio. If the ratio is separate from stops, then it becomes pretty complicated UI-wise - too many ways to setup a cone operation. With current approach you can easily cut cone left-right-left-right without having to move back - just by using reverse button. You also don't have to wrestle with calculations much. I think the ultimate answer here would be https://kachurovskiy.github.io/lathecode/ which (once operational) will take all the guesswork out.

OK, makes sense. I still feel it's a bit counterintuitive, but I will give it a try with the back and forth.



>>  Lathe running in reverse behavior or on screen warning? As of now, the NanoEls just does… nothing.

> You mean in automatic mode? Yes, an on-screen warning with a beep would be nice. Hopefully not a pressing issue :)

Not pressing, but I have really wondered for 2-3 minutes what was wrong with the H4 while testing yesterday... 🤦🏻‍♂️



>>  Facing operations: facing Up to Down as well as Down to Up (or in reverse). For example, when using a boring bar to face a relief. I ended up doing an internal turning operation instead.

> I believe this is supported already, just set the negative pitch.

Ah ok. I will test that.



>>  jump to next operation: map one button to jump to the next operation. Could be very useful if the soft stops were set hastily and the tool doesn't engage the workpiece yet.

> This is an interesting idea and probably not hard to implement :) I'll see if it works.

👍👍



>>  Quick Jog / precision mode toggle? (going through 1mm 0,1mm 0,01mm 0,001mm every time gets old quick 😃 )

> No idea how that would work UI-wise.

Maybe the toggle could be long press ![239697700-52ce78e4-4202-4642-9433-e61ca39de9d5](https://github.com/kachurovskiy/nanoels/assets/122809499/a0b32342-a442-47d3-afef-f6c0ea88f326)



>>  Set current X/Y position to XXX (as opposed to reset current position to 0 only)

> This is high on my priority list, it's needed for GCode zeroing.

Yeah! 👍👍



>>  Possibility to disable/enable all soft stops (for example to move the carriage out of the way to measure the part, without losing the stop positions)

> I was thinking about it but don't see a practical necessity for it yet, setting a right stop in a slightly different place after measuring doesn't seem to affect any of the operations AFAIU?

Actually, it could, if you have your DOC and number of passes dialed in just right. First UI solution that comes to mind: temporarily disable each limit, by long pressing the corresponding limit button?

**Comment by fmw626** [comment]
@kachurovskiy 
While doing further test today, I found 2 more issues:

Position drift: I did a test with a dial indicator fixed to the lathe bed, indicating against the carriage:
![IMG_7394](https://github.com/kachurovskiy/nanoels/assets/122809499/7d09a175-0a35-4c26-a495-d553c72f0c7a)
I then went back and forth on the Z axis: 100x 1 mm (50x left, 50x right). After all these operations, the position of the carriage drifted 0,1 mm from the initial position. My guess is it's a cumulative rounding error? Is there a way to fix this? This is something that could cause serious tolerance issues on complicated parts.

The second issue I found during the same test, is the behavior of the backlash compensation when the desired movement is of a smaller distance than the amount of backlash set for that axis: It doesn't move, and therefore the position is wrong. 
My ideas to fix this issue:
            - On-screen warning that we are within the backlash zone and the position is not accurate
            - Automatic move out further than the pre-defined backlash distance, then back in. 
               e.g.: backlash is 1 mm. Desired movement: 0,2 mm to the right. Actual movement: 1,5 mm to the right, then 1,3 mm to the left.

**Reply by kachurovskiy** [reply]
> the position of the carriage drifted 0,1 mm from the initial position

There's something interesting going on there, I couldn't reproduce this on 1mm or 0.1mm but when I started to move back and forth quickly on 1mm it happened. Need to investigate.

> behavior of the backlash compensation when the desired movement is of a smaller distance than the amount of backlash set for that axis: It doesn't move, and therefore the position is wrong.

I was unable to immediately reproduce this, if you could share the exact steps or a quick video that would really help! Thanks a lot.

**Reply by fmw626** [reply]
Hi @kachurovskiy 

>>behavior of the backlash compensation when the desired movement is of a smaller distance than the amount of backlash set for that axis: It doesn't move, and therefore the position is wrong.

>I was unable to immediately reproduce this, if you could share the exact steps or a quick video that would really help! Thanks a lot.

So while taking a video to explain the issue better, I came to realize that I interpreted the behavior wrong. What is happening is twofold:
- The backlash compensation travel speed changes with the selected step speed. At 0,01 mm it is slow, but at 0,001 mm it is really slow. I thought the H4 was already moving the carriage according to my input, but it was still taking up the backlash, thus i came to the wrong conclusions. Does it need to be that slow with 0,001 mm?
- 0,001 mm is really, really small. The amount of backlash I set is about 0,3 mm, so about 300x the step resolution. I realized it might be futile to dial in the backlash amount to that precise of a measure. No way it is 100% consistant along the entire leadscrew...

**Reply by fmw626** [reply]
>>the position of the carriage drifted 0,1 mm from the initial position

>There's something interesting going on there, I couldn't reproduce this on 1mm or 0.1mm but when I started to move back and forth quickly on 1mm it happened. Need to investigate.

Here are 2 videos showing the behavior in detail. One is 100(ish) travels at 1 mm, the other is 100(ish) travels at 10 mm. Notice how both drift, but also not the same amount.


https://github.com/kachurovskiy/nanoels/assets/122809499/9f48312c-6702-431a-816b-f95e1e3e2c12


https://github.com/kachurovskiy/nanoels/assets/122809499/7ab3d504-733e-4cda-8f6f-fa00b0542f1e

**Reply by kachurovskiy** [reply]
Thank you so much for this investigation. I hope your lead screw was well oiled during the test 😀 

Checking the code, I see that we should be issuing steps to go to the same position over and over. However, driver seems to require a bit of time after a direction change that we might not have provided before - https://www.omc-stepperonline.com/download/CL57T_V4.0.pdf page 1 "Minimal direction setup" is 2us. I've added this delay in `setDir()` in my local code, will be testing it.

```
void setDir(Axis* a, bool dir) {
  // Start slow if direction changed.
  if (a->direction != dir || !a->directionInitialized) {
    a->speed = a->speedStart;
    a->direction = dir;
    a->directionInitialized = true;
    DWRITE(a->dir, dir ^ a->invertStepper);
    delayMicroseconds(DIRECTION_SETUP_DELAY_US);
  }
}
```

**Reply by kachurovskiy** [reply]
The fix is now in the main repo - https://github.com/kachurovskiy/nanoels/commit/2134a98fb0e6f22e5bdb20405a8e6dc0a77ec2c3 - hope it helps.

**Comment by fmw626** [comment]
Hi @kachurovskiy 

I took the time to do some more testing today, and wanted to share an update on all the points I raised:

- X-axis stall

     Still happening. Even if it’s a shot in the dark, I will try messing with the stepper driver settings today to see if it changes anything. I don’t understand why the issue only occurs on my setup. 



- Unexpected movement in the wrong direction after cone cut

     Fixed! 👍


- Position drift after many operations

     Fixed! 💪💪


- Confusing backlash behavior at micron step and tedious step resolution changes

     Fixed by ditching micron


- X movement, not following steps when long pressing
	
     Fixed! 👍


- As you have suggested [here](https://github.com/kachurovskiy/nanoels/discussions/121#discussioncomment-6083895), I think that arrow symbols would be more intuitive than "LRUD"

     Implemented! 👍


- Special space-saving mm character:

     Implemented! 👍


- Facing UP to Down

     Like you said, it already works by setting the pitch to negative. My bad.


- X position disappearing when X=0

     Half-fixed via https://github.com/kachurovskiy/nanoels/discussions/120#discussioncomment-6082502, but behavior not satisfactory: When X=0, “Position” appears before Z, and there is no space between X and Z, and also no unit. It would make more sense if the display showed “0mm 0mm” instead of “Position 00”.


https://github.com/kachurovskiy/nanoels/assets/122809499/e59c1cb8-0145-4773-a624-49025c1057ef


- Cone mode behavior

     While I am still not convinced this is the most intuitive implementation, I did some tests with changing direction like you suggested, and it works quite well. I think what bugs me about this behavior is that it cannot do multiple operations automatically, like in the turning and facing modes.


Not implemented/addressed yet

- Set current X/Y position to XXX
- Skip to next operation by pressing play
- DOC instead of number of passes
- Temporarily disable soft limits (by long pressing the limit button?)
- Cutoff: Why ask the number of passes? Shouldn't it always be only one pass?
- Lathe running in reverse behavior or on screen warning? As of now, the NanoEls just does… nothing.

**Reply by kachurovskiy** [reply]
> X-axis stall

Please give me a bit more time, I'll try to implement some counter-measures (even though I can't reproduce the bug). If in your day-to-day work you notice a short and repeatable path to reproduce this bug, please do let me know.

> X position disappearing when X=0

I'll add a constant to control this 👍

> Cone mode behavior

I agree with you here, multi-pass cone is needed in addition to what we have today - though it might take a while before it's implemented.

> Cutoff: Why ask the number of passes? Shouldn't it always be only one pass?

In some cases the cutter can get clogged so sometimes pulling out to clear the chips could be useful. I've seen it on narrow tools and deep cuts, other ELS also have this feature I think. I hope it's not in the way as you can press `1` during setup and have the single-pass logic.

For the rest of your items, the feedback is taken but I plan to implement "Set current X/Y position to XXX" and collect more feedback on others items. Hope that's agreeable 😀

Have a good weekend!

**Reply by fmw626** [reply]
Hey @kachurovskiy 

>If in your day-to-day work you notice a short and repeatable path to reproduce this bug, please do let me know.

Of course! 🙏

> I agree with you here, multi-pass cone is needed in addition to what we have today - though it might take a while before it's implemented.

Perfect! Not urgent, the back and forth you suggested does the trick for now.

> I hope it's not in the way as you can press 1 during setup and have the single-pass logic.

Ah ok, I understand. It's definitely not in the way. Maybe this would work well with "Skip to next operation by pressing play", so it's quick to retract the blade and start another pass if it starts to jam? Just an idea.

>... and collect more feedback on others items. Hope that's agreeable 😀

By all means! I'd also like to hear what other users say.

Good weekend to you!

**Reply by kachurovskiy** [reply]
In the latest software version I've implemented:

- a speculative fix for the motor stall
- setting axis 0 at a given distance ahead
- constants allowing to always show X position
- constants allowing to show axis name before position value

Please let me know if it works for you. Cheers!
